File: src/domain/services/applicationService_test.go
package services

import (
	"fmt"
	"testing"

	"Theatrum/domain/models"
)

// appTestStorage implements repositories.StoragePort with configurable behavior.
type appTestStorage struct {
	searchFilesFn   func(pattern string, extensions []string) ([]string, []map[string]string, error)
	readFileData    []byte
	readFileErr     error
	writeFileErr    error
	deleteFileErr   error
}

func (m *appTestStorage) ReadFile(path string) ([]byte, error) {
	return m.readFileData, m.readFileErr
}
func (m *appTestStorage) WriteFile(path string, data []byte) error {
	return m.writeFileErr
}
func (m *appTestStorage) DeleteFile(path string) error {
	return m.deleteFileErr
}
func (m *appTestStorage) ListFiles(pattern string) ([]string, error) {
	return nil, nil
}
func (m *appTestStorage) GetFileSize(path string) (int64, error) {
	return 0, nil
}
func (m *appTestStorage) SearchFiles(pattern string, extensions []string) ([]string, []map[string]string, error) {
	if m.searchFilesFn != nil {
		return m.searchFilesFn(pattern, extensions)
	}
	return nil, nil, nil
}

func TestApplicationService_GetApplication(t *testing.T) {
	app := &models.Application{PublicPath: "http://localhost:8080"}
	server := &models.Server{HTTPPort: 8080}
	channels := map[string]models.Stream{}
	svc := NewApplicationService(app, server, &channels, &appTestStorage{}, NewPathTemplateService())

	got := svc.GetApplication()
	if got != app {
		t.Error("expected same application pointer")
	}
	if got.PublicPath != "http://localhost:8080" {
		t.Errorf("expected PublicPath http://localhost:8080, got %q", got.PublicPath)
	}
}

func TestApplicationService_GetServer(t *testing.T) {
	app := &models.Application{}
	server := &models.Server{HTTPPort: 8080, RTMPPort: 1935}
	channels := map[string]models.Stream{}
	svc := NewApplicationService(app, server, &channels, &appTestStorage{}, NewPathTemplateService())

	got := svc.GetServer()
	if got != server {
		t.Error("expected same server pointer")
	}
	if got.HTTPPort != 8080 {
		t.Errorf("expected HTTPPort 8080, got %d", got.HTTPPort)
	}
}

func TestApplicationService_GetChannels(t *testing.T) {
	app := &models.Application{}
	server := &models.Server{}
	channels := map[string]models.Stream{
		"/user/{username}": {Type: models.StreamTypeLive},
	}
	svc := NewApplicationService(app, server, &channels, &appTestStorage{}, NewPathTemplateService())

	got := svc.GetChannels()
	if len(*got) != 1 {
		t.Errorf("expected 1 channel, got %d", len(*got))
	}
}

func TestApplicationService_GetChannel(t *testing.T) {
	app := &models.Application{}
	server := &models.Server{}
	channels := map[string]models.Stream{
		"/user/{username}": {Type: models.StreamTypeLive, Path: "live/{username}"},
	}
	svc := NewApplicationService(app, server, &channels, &appTestStorage{}, NewPathTemplateService())

	t.Run("found", func(t *testing.T) {
		ch, err := svc.GetChannel("/user/{username}")
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if ch.Type != models.StreamTypeLive {
			t.Errorf("expected StreamTypeLive, got %s", ch.Type)
		}
	})

	t.Run("not found", func(t *testing.T) {
		_, err := svc.GetChannel("/nonexistent")
		if err == nil {
			t.Error("expected error for nonexistent channel")
		}
	})
}

func TestApplicationService_BuildAllStreamsPlaylist(t *testing.T) {
	t.Run("disabled returns error", func(t *testing.T) {
		app := &models.Application{AllStreamsPlaylist: models.AllStreamsPlaylist{Enabled: false}}
		channels := map[string]models.Stream{}
		svc := NewApplicationService(app, &models.Server{}, &channels, &appTestStorage{}, NewPathTemplateService())

		_, err := svc.BuildAllStreamsPlaylist()
		if err == nil {
			t.Error("expected error when disabled")
		}
	})

	t.Run("enabled builds M3U8", func(t *testing.T) {
		app := &models.Application{
			PublicPath: "http://localhost:8080",
			AllStreamsPlaylist: models.AllStreamsPlaylist{
				Enabled: true,
				Path:    "/all.m3u8",
			},
		}
		channels := map[string]models.Stream{
			"/user/{username}": {
				Type: models.StreamTypeVideoEncoded,
				Path: "videos/{username}",
			},
		}
		storage := &appTestStorage{
			searchFilesFn: func(pattern string, extensions []string) ([]string, []map[string]string, error) {
				return []string{"videos/alice/master.m3u8"}, []map[string]string{{"username": "alice"}}, nil
			},
		}
		svc := NewApplicationService(app, &models.Server{}, &channels, storage, NewPathTemplateService())

		result, err := svc.BuildAllStreamsPlaylist()
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if result == "" {
			t.Error("expected non-empty playlist")
		}
		if !contains(result, "#EXTM3U") {
			t.Error("expected M3U8 header")
		}
		if !contains(result, "master.m3u8") {
			t.Error("expected master.m3u8 reference in playlist")
		}
	})

	t.Run("storage search error is handled gracefully", func(t *testing.T) {
		app := &models.Application{
			PublicPath: "http://localhost:8080",
			AllStreamsPlaylist: models.AllStreamsPlaylist{Enabled: true, Path: "/all.m3u8"},
		}
		channels := map[string]models.Stream{
			"/user/{username}": {Type: models.StreamTypeVideoEncoded, Path: "videos/{username}"},
		}
		storage := &appTestStorage{
			searchFilesFn: func(pattern string, extensions []string) ([]string, []map[string]string, error) {
				return nil, nil, fmt.Errorf("disk error")
			},
		}
		svc := NewApplicationService(app, &models.Server{}, &channels, storage, NewPathTemplateService())

		result, err := svc.BuildAllStreamsPlaylist()
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		// Should still return valid M3U8 header even with no streams
		if !contains(result, "#EXTM3U") {
			t.Error("expected M3U8 header")
		}
	})
}

func contains(s, substr string) bool {
	return len(s) >= len(substr) && containsSubstring(s, substr)
}

func containsSubstring(s, sub string) bool {
	for i := 0; i <= len(s)-len(sub); i++ {
		if s[i:i+len(sub)] == sub {
			return true
		}
	}
	return false
}

File: src/test/java/com/fjourdren/theatrum/domain/service/ApplicationServiceTest.java
package com.fjourdren.theatrum.domain.service;

import com.fjourdren.theatrum.application.port.out.StoragePort;
import com.fjourdren.theatrum.domain.model.AllStreamsPlaylist;
import com.fjourdren.theatrum.domain.model.AppPaths;
import com.fjourdren.theatrum.domain.model.Application;
import com.fjourdren.theatrum.domain.model.FileMatch;
import com.fjourdren.theatrum.domain.model.LoadedConfiguration;
import com.fjourdren.theatrum.domain.model.Server;
import com.fjourdren.theatrum.domain.model.Stream;
import com.fjourdren.theatrum.domain.model.StreamType;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.io.IOException;
import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class ApplicationServiceTest {

    private static final AppPaths PATHS = new AppPaths(Path.of("data"), Path.of("frontend"));

    @Mock
    private StoragePort storage;

    private static Application application(String publicPath, boolean playlistEnabled) {
        return new Application(publicPath, new AllStreamsPlaylist(playlistEnabled, "/all.m3u8"));
    }

    private ApplicationService service(Application app, Server server, Map<String, Stream> channels) {
        return new ApplicationService(new LoadedConfiguration(app, server, channels), storage,
                new PathTemplateService(), PATHS);
    }

    @Test
    void getApplicationReturnsTheConfiguredApplication() {
        Application app = application("http://localhost:8080", false);

        Application got = service(app, new Server(8080, 0, null), Map.of()).getApplication();

        assertThat(got).isSameAs(app);
        assertThat(got.publicPath()).isEqualTo("http://localhost:8080");
    }

    @Test
    void getServerReturnsTheConfiguredServer() {
        Server server = new Server(8080, 1935, null);

        Server got = service(application("", false), server, Map.of()).getServer();

        assertThat(got).isSameAs(server);
        assertThat(got.httpPort()).isEqualTo(8080);
    }

    @Test
    void getChannelsReturnsEveryChannel() {
        Map<String, Stream> channels = Map.of(
                "/user/{username}", Stream.builder().type(StreamType.LIVE).build());

        assertThat(service(application("", false), new Server(0, 0, null), channels).getChannels())
                .hasSize(1);
    }

    @Nested
    class GetChannel {

        private final Map<String, Stream> channels = Map.of("/user/{username}",
                Stream.builder().type(StreamType.LIVE).path("live/{username}").build());

        @Test
        void found() {
            Stream channel = service(application("", false), new Server(0, 0, null), channels)
                    .getChannel("/user/{username}");

            assertThat(channel.type()).isEqualTo(StreamType.LIVE);
        }

        @Test
        void notFound() {
            ApplicationService service = service(application("", false), new Server(0, 0, null), channels);

            assertThatThrownBy(() -> service.getChannel("/nonexistent"))
                    .isInstanceOf(IllegalArgumentException.class)
                    .hasMessage("channel not found: /nonexistent");
        }
    }

    @Nested
    class BuildAllStreamsPlaylist {

        @Test
        void disabledThrows() {
            ApplicationService service =
                    service(application("", false), new Server(0, 0, null), Map.of());

            assertThatThrownBy(service::buildAllStreamsPlaylist)
                    .isInstanceOf(IllegalStateException.class)
                    .hasMessage("all streams playlist is not enabled");
        }

        @Test
        void enabledBuildsM3u8() throws IOException {
            Map<String, Stream> channels = new LinkedHashMap<>();
            channels.put("/user/{username}", Stream.builder()
                    .type(StreamType.VIDEO_ENCODED).path("videos/{username}").build());
            when(storage.searchFiles(anyString(), any())).thenReturn(List.of(
                    new FileMatch(Path.of("videos/alice/master.m3u8"), Map.of("username", "alice"))));

            String result = service(application("http://localhost:8080", true),
                    new Server(0, 0, null), channels).buildAllStreamsPlaylist();

            assertThat(result)
                    .contains("#EXTM3U")
                    .contains("#EXT-X-VERSION:3")
                    .contains("#EXT-X-STREAM-INF:BANDWIDTH=0")
                    .contains("http://localhost:8080/user/alice/master.m3u8");
        }

        @Test
        void searchesUnderTheVideoDirectory() throws IOException {
            Map<String, Stream> channels = Map.of("/user/{username}", Stream.builder()
                    .type(StreamType.VIDEO_ENCODED).path("videos/{username}").build());
            when(storage.searchFiles(anyString(), any())).thenReturn(List.of());

            service(application("http://localhost:8080", true), new Server(0, 0, null), channels)
                    .buildAllStreamsPlaylist();

            org.mockito.Mockito.verify(storage).searchFiles(
                    Path.of("data", "videos/{username}", "master.m3u8").toString(),
                    com.fjourdren.theatrum.domain.constant.VideoConstants.VALID_MASTER_PLAYLIST_EXTENSIONS);
        }

        @Test
        void storageSearchErrorIsHandledGracefully() throws IOException {
            Map<String, Stream> channels = Map.of("/user/{username}", Stream.builder()
                    .type(StreamType.VIDEO_ENCODED).path("videos/{username}").build());
            when(storage.searchFiles(anyString(), any())).thenThrow(new IOException("disk error"));

            String result = service(application("http://localhost:8080", true),
                    new Server(0, 0, null), channels).buildAllStreamsPlaylist();

            assertThat(result).contains("#EXTM3U").doesNotContain("master.m3u8");
        }

        @Test
        void templateFailureSkipsTheEntry() throws IOException {
            Map<String, Stream> channels = Map.of("/user/{username}", Stream.builder()
                    .type(StreamType.VIDEO_ENCODED).path("videos/{username}").build());
            // "../etc" cannot be sanitized, so placeholder replacement fails for this match.
            when(storage.searchFiles(anyString(), any())).thenReturn(List.of(
                    new FileMatch(Path.of("videos/x/master.m3u8"), Map.of("username", "../etc"))));

            String result = service(application("http://localhost:8080", true),
                    new Server(0, 0, null), channels).buildAllStreamsPlaylist();

            assertThat(result).contains("#EXTM3U").doesNotContain("master.m3u8");
        }
    }
}
